File: rust/examples/channel-options-example.rs
use helius_laserstream::{subscribe, ChannelOptions, LaserstreamConfig, SubscribeRequest, SubscribeRequestFilterSlots};
use futures::StreamExt;
use std::collections::HashMap;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let endpoint = std::env::var("LASERSTREAM_PRODUCTION_ENDPOINT")
        .unwrap_or_else(|_| "".to_string());
    let api_key = std::env::var("LASERSTREAM_PRODUCTION_API_KEY")
        .expect("LASERSTREAM_PRODUCTION_API_KEY environment variable must be set");

    // Create custom channel options
    let channel_options = ChannelOptions {
        // Connection timeouts
        connect_timeout_secs: Some(20),              // 20 seconds instead of default 10
        timeout_secs: Some(60),                      // 60 seconds instead of default 30
        
        // Message size limits
        max_decoding_message_size: Some(2_000_000_000),  // 2GB instead of default 1GB
        max_encoding_message_size: Some(64_000_000),     // 64MB instead of default 32MB
        
        // Keep-alive settings
        http2_keep_alive_interval_secs: Some(15),   // 15 seconds instead of default 30
        keep_alive_timeout_secs: Some(10),           // 10 seconds instead of default 5
        keep_alive_while_idle: Some(true),
        
        // Window sizes for flow control
        initial_stream_window_size: Some(8 * 1024 * 1024),      // 8MB instead of default 4MB
        initial_connection_window_size: Some(16 * 1024 * 1024), // 16MB instead of default 8MB
        
        // Performance options
        http2_adaptive_window: Some(true),
        tcp_nodelay: Some(true),
        tcp_keepalive_secs: Some(30),               // 30 seconds instead of default 60
        buffer_size: Some(128 * 1024),               // 128KB instead of default 64KB
    };

    // Create config with custom channel options
    let config = LaserstreamConfig::new(endpoint, api_key)
        .with_max_reconnect_attempts(5)
        .with_channel_options(channel_options);

    // Create a simple slot subscription request
    let mut slots_filter = HashMap::new();
    slots_filter.insert(
        "client".to_string(),
        SubscribeRequestFilterSlots::default(),
    );

    let request = SubscribeRequest {
        slots: slots_filter,
        ..Default::default()
    };

    println!("Subscribing with custom channel options...");
    println!("- Connect timeout: 20s");
    println!("- Max receive message size: 2GB");
    println!("- Keep-alive interval: 15s");
    println!("- Initial stream window: 8MB");

    let (stream, _handle) = subscribe(config, request);
    let mut stream = Box::pin(stream);

    while let Some(update) = stream.next().await {
        match update {
            Ok(update) => {
                println!("Received update: {:?}", update);
            }
            Err(e) => {
                eprintln!("Stream error: {:?}", e);
                break;
            }
        }
    }

    Ok(())
}

File: javascript/examples/channel-options-example.ts
import { subscribe, CommitmentLevel, SubscribeUpdate, StreamHandle, LaserstreamConfig, ChannelOptions } from '../client';

const credentials = require('../test-config');

async function main() {
  console.log('⚙️  Laserstream Channel Options Example');

  // Custom channel options
  const channelOptions: ChannelOptions = {
    // Connection timeouts
    connectTimeoutSecs: 20,              // 20 seconds instead of default 10
    timeoutSecs: 60,                     // 60 seconds instead of default 30
    
    // Message size limits
    maxDecodingMessageSize: 2000000000,  // 2GB instead of default 1GB
    maxEncodingMessageSize: 64000000,    // 64MB instead of default 32MB
    
    // Keep-alive settings
    http2KeepAliveIntervalSecs: 15,     // 15 seconds instead of default 30
    keepAliveTimeoutSecs: 10,           // 10 seconds instead of default 5
    keepAliveWhileIdle: true,
    
    // Window sizes for flow control
    initialStreamWindowSize: 8388608,    // 8MB instead of default 4MB
    initialConnectionWindowSize: 16777216, // 16MB instead of default 8MB
    
    // Performance options
    http2AdaptiveWindow: true,
    tcpNodelay: true,
    tcpKeepaliveSecs: 30,               // 30 seconds instead of default 60
    bufferSize: 131072,                 // 128KB instead of default 64KB
  };

  // Configuration with custom channel options
  const config: LaserstreamConfig = {
    apiKey: credentials.laserstreamProduction.apiKey,
    endpoint: credentials.laserstreamProduction.endpoint,
    maxReconnectAttempts: 5,
    channelOptions: channelOptions
  };

  // Subscription request
  const request = {
    slots: {
      client: {
        filterByCommitment: true,
      }
    },
    commitment: CommitmentLevel.CONFIRMED
  };

  try {
    const stream: StreamHandle = await subscribe(
      config,
      request,
      // Data callback
      async (update: SubscribeUpdate) => {
        if (update.slot) {
          console.log(`🎰 Slot update: ${update.slot.slot}`);
        } else {
          console.log('📦 Received update:', update);
        }
      },
      // Error callback
      async (error: Error) => {
        console.error('❌ Stream error:', error);
      }
    );

    console.log(`✅ Stream started with ID: ${stream.id}`);

    // Handle graceful shutdown
    process.on('SIGINT', async () => {
      console.log('\n🛑 Shutting down...');
      stream.cancel();
      process.exit(0);
    });

  } catch (error) {
    console.error('❌ Failed to subscribe:', error);
    process.exit(1);
  }
}

main().catch(console.error);